### PR TITLE
next: Enhance the variation read and display

### PR DESCRIPTION
**Pull request by zsalch**
1. Support more sgf variation format and enhance the variation display.
2. New move number in the branch, the option is 'new-move-number-in-branch', default on.

### CODE REVIEW

**Comment by featurecat** [comment]
Even though there has been no review I am tempted to merge this. Thoughts? @OlivierBlanvillain

**Comment by OlivierBlanvillain** [comment]
The diff is +347-76, I think it deserves a review ^^'

**Comment by zsalch** [comment]
Here are some tests: [#2](https://github.com/zsalch/lizzie/issues/2) [#9](https://github.com/zsalch/lizzie/issues/9)

File: src/main/java/featurecat/lizzie/Config.java [hunk 1]
```diff
@@ -14,6 +14,7 @@
 public class Config {
 
   public boolean showMoveNumber = false;
+  public boolean newMoveNubmerInBranch = true;
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/Config.java]: Fixed.

File: src/main/java/featurecat/lizzie/rules/BoardData.java [hunk 2]
```diff
@@ -6,9 +6,11 @@
 
 public class BoardData {
   public int moveNumber;
+  public int moveMNNumber;
   public Optional<int[]> lastMove;
   public int[] moveNumberList;
   public boolean blackToPlay;
+  public boolean dummy;
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/BoardData.java]: There are some sgf file contains the AB/AW/AE attributes. For these attributes, it is better to create a dummy BoardData.

File: src/main/java/featurecat/lizzie/rules/BoardData.java [hunk 3]
```diff
@@ -6,9 +6,11 @@
 
 public class BoardData {
   public int moveNumber;
+  public int moveMNNumber;
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/BoardData.java]: In current code, the moveNumber actual keep the BoardData sequence number.
Many logic depend on the moveNumber.
The moveMNNumber for display move number, such as 'MN' attribute, move number in the branch.

File: src/main/java/featurecat/lizzie/rules/BoardHistoryNode.java [hunk 4]
```diff
@@ -306,6 +364,69 @@ public int findIndexOfNode(BoardHistoryNode childNode) {
     return -1;
   }
 
+  /**
+   * Given a child node, find the index of that child node in it's parent
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/BoardHistoryNode.java]: Fixed.

File: src/main/java/featurecat/lizzie/rules/BoardHistoryNode.java [hunk 5]
```diff
@@ -306,6 +364,69 @@ public int findIndexOfNode(BoardHistoryNode childNode) {
...
+   *
+   * @return index of child node, -1 if child node not a child of parent
+   */
+  public int findIndexOfNode(BoardHistoryNode childNode, boolean allSub) {
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/BoardHistoryNode.java]: Fixed.

File: src/main/java/featurecat/lizzie/rules/BoardHistoryNode.java [hunk 6]
```diff
@@ -306,6 +364,69 @@ public int findIndexOfNode(BoardHistoryNode childNode) {
...
+        }
+        move++;
+        node = node.get().next();
+      }
+    }
+    return 0;
+  }
+
+  /**
+   * The move number of that node in it's branch
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/BoardHistoryNode.java]: Fixed.

File: src/main/java/featurecat/lizzie/gui/VariationTree.java [hunk 7]
```diff
@@ -58,60 +60,74 @@ public void drawTree(
...
-    }
-    if (startNode.previous().isPresent()) {
-      g.fillOval(curposx, posy, DOT_DIAM, DOT_DIAM);
-      g.setColor(Color.BLACK);
-      g.drawOval(curposx, posy, DOT_DIAM, DOT_DIAM);
-    } else {
-      g.fillRect(curposx, posy, DOT_DIAM, DOT_DIAM);
-      g.setColor(Color.BLACK);
-      g.drawRect(curposx, posy, DOT_DIAM, DOT_DIAM);
+    if (curposx > minposx && posy > 0) {
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/gui/VariationTree.java]: In order to show the variation of those top node outside the display area, must check previous node to top, these node's posy will be <= 0.

File: src/main/java/featurecat/lizzie/gui/VariationTree.java [hunk 8]
```diff
@@ -166,6 +182,56 @@ public void draw(Graphics2D g, int posx, int posy, int width, int height) {
...
+  private int getCurLane(
+      BoardHistoryNode start, BoardHistoryNode curMove, int curposy, int maxy, boolean isMain) {
+    BoardHistoryNode next = start;
+    int nexty = curposy;
+    while (next.next().isPresent() && nexty + YSPACING < maxy) {
+      nexty += YSPACING;
+      next = next.next().get();
+    }
+    BoardHistoryNode end = isMain ? null : start;
+    while (next.previous().isPresent() && next != end) {
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/gui/VariationTree.java]: Fixed.

File: src/main/java/featurecat/lizzie/gui/VariationTree.java [hunk 9]
```diff
@@ -166,6 +182,56 @@ public void draw(Graphics2D g, int posx, int posy, int width, int height) {
       node = node.previous().get();
       curposy -= YSPACING;
     }
-    drawTree(g, posx + xoffset, curposy, 0, posy + height, node, 0, true);
+    laneCount = 0;
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/gui/VariationTree.java]: Fixed.

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 10]
```diff
@@ -219,20 +239,35 @@ private static boolean parse(String value) {
...
-                Lizzie.board.moveNumber(Integer.parseInt(tagContent));
-              } else if ("AE".equals(tag)) {
+              if ("AE".equals(tag)) {
                 // remove a stone
-                if (addPassForAwAb) {
-                  Lizzie.board.pass(tag.equals("AB") ? Stone.BLACK : Stone.WHITE);
-                  addPassForAwAb = false;
+                if (addPassForMove) {
+                  // Save the step count
+                  subTreeStepMap.put(subTreeDepth, subTreeStepMap.get(subTreeDepth) + 1);
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: Yes, I agree. However, it takes a lot of time, and maybe do it in the future if have time.

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 11]
```diff
@@ -558,4 +595,18 @@ public static String nodeString(String key, String value) {
     }
     return sb.toString();
   }
+
+  private static void processProperties(String tag, String tagContent) {
+    Lizzie.board.addNodeProperty(tag, tagContent);
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: Fixed.

File: src/main/java/featurecat/lizzie/rules/BoardData.java [hunk 12]
```diff
@@ -72,6 +76,11 @@ public static BoardData empty(int size) {
    */
   public void addProperty(String key, String value) {
     SGFParser.addProperty(properties, key, value);
+    if ("N".equals(key) && comment.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (zsalch) [reply src/main/java/featurecat/lizzie/rules/BoardData.java]: Yes, I think these code needs to be refactored and optimized in the next version.

**Comment by featurecat** [comment]
Sorry for not enough time to review the changes further, I have to hurry and release 0.6. I agree about no SGF parsing outside of the SGF parser classes.

File: src/main/java/featurecat/lizzie/rules/SGFParser.java [hunk 13]
```diff
@@ -558,4 +595,18 @@ public static String nodeString(String key, String value) {
...
+    if ("MN".equals(tag)) {
+      Lizzie.board.moveNumber(Integer.parseInt(tagContent));
+    }
+  }
+
+  private static void processPendingPros(Map<String, String> props) {
+    if (props.size() > 0) {
+      props.forEach((key, value) -> processProperties(key, value));
+      props = new HashMap<String, String>();
```
[the remark this answers is not in the document]
Comment (OlivierBlanvillain) [reply src/main/java/featurecat/lizzie/rules/SGFParser.java]: I think this doesn't clear anything. It creates a new HashMap and update the *local* variable props. What you want is `props.clear()`.

File: src/main/java/featurecat/lizzie/rules/BoardData.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (OlivierBlanvillain) [reply src/main/java/featurecat/lizzie/rules/BoardData.java]: I'm not really happy with keeping all this state around, its quite hard to follow to be honest. Do you think we could we instead recompute `moveNumber` on the fly? I also think the naming doesn't help here. Maybe we should rename these things as follows:

- moveNumber → depth
- moveMNNumber → numberlabel

WDYT?

File: src/main/java/featurecat/lizzie/rules/BoardData.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (OlivierBlanvillain) [reply src/main/java/featurecat/lizzie/rules/BoardData.java]: Still, I don't understand. Why do we need dummy BoardData?
